Orange/data/table.py

```diff
@@ -33,146 +28,176 @@ def get_sample_datasets_dir():
+        Exports a numpy matrix. The order is always X, Y, meta.
+        The columns are in the same order as in
+        Table.columns, Table.columns_X, Table.columns_Y, Table.columns_meta
+        If writable == False (default), the numpy writable flag is set to false.
+            This means write operations on this array will loudly fail. 
         """
-        return Columns(self.domain)
+        # TODO: only return numeric values here, need to transform
+        cols = []
+        cols += self.columns_X if X else []
+        cols += self.columns_Y if Y else []
+        cols += self.columns_meta if meta else []
+        res = self[cols].values
+        res.setflags(write=writable)
+        return res
 
-    _next_instance_id = 0
-    _next_instance_lock = Lock()
+    @property
+    def X(self):
+        """
+        Return a read-only numpy matrix of X.
+        Always a two-dimensional matrix, even for a single column.
+        The columns are in the same order as returned by Table.columns_X.
+        """
+        return self._to_numpy(X=True)
 
     @property
     def Y(self):
-        if self._Y.shape[1] == 1:
-            return self._Y[:, 0]
-        return self._Y
+        """
+        Return a read-only numpy array/matrix of Y.
+        If there is only one column, returns a one-dimensional array,
+            otherwise, returns a two-dimensional matrix whose columns
+            are in the same order as returned by Table.columns_Y.
+        """
+        res = self._to_numpy(Y=True)
+        # TODO: comment on whether this separation is okay please
+        return res.flatten() if len(self.columns_Y) else res
+
+    @property
+    def metas(self):
+        """
+        Return a read-only numpy matrix of meta attributes.
+        Always a two-dimensional matrix, even for a single column.
+        The columns are in the same order as returned by Table.columns_metas.
+        """
+        return self._to_numpy(meta=True)
+
+    @property
+    def weights(self):
+        # TODO: do we switch to .weights and deprecate .W or keep .W?
+        """Get the weights as a numpy array."""
+        return self[Table._WEIGHTS_COLUMN].values
 
-    @Y.setter
-    def Y(self, value):
-        if len(value.shape) == 1:
-            value = value[:, None]
-        self._Y = value
+    @property
+    def W(self):
+        return self.weights
+
+    def set_role(self, column_names, column_roles):
+        """
+        Sets the role a column (or multiple columns) takes in this table.
+        """
+        if isinstance(column_names, str):
+            column_names = [column_names]
+        if isinstance(column_roles, str):
+            column_roles = [column_roles]
+        for n, rstr in zip(column_names, column_roles):
+            r = Role.from_string(rstr)
+            self._columns_X.discard(n)
+            self._columns_Y.discard(n)
+            self._columns_meta.discard(n)
+            if r == Role.x:
+                self._columns_X.add(n)
```

Comment:
Since a var supposedly can't be X, Y, and meta simultaneously, how about having just a single _var_roles that holds the roles of each var.
If _var_roles is an ordereddict, you can simply do:
```py
    self._var_roles[n] = r
```